w3af/plugins/crawl/soap.py

```diff
@@ -0,0 +1,168 @@
+                "specification in WSDL format. The WSDL specification gives us info"
+                "what methods are available and how to use them. It's a road map"
+                "for SOAP client."
+            ),
+            _type=option_types.STRING,
+        )
+        option_list.add(option)
+
+        option = opt_factory(
+            'verbosity_level',
+            default_value=self.verbosity_level,
+            desc='Verbosity level',
+            help=(
+                "You can specify number as below:\n"
+                "0: No verbosity. Plugin will only detect new URLs.\n"
+                "1: Plugin will detect URLs and it will add info to knowledge base"
+                "about possible operations on given URL.\n"
+                "2: Plugin will detect URLs and dump all info from WSDL spec."
+            ),
+            _type=option_types.INT,
+        )
+        option_list.add(option)
+
+        return option_list
+
+    def set_options(self, options_list):
+        super(soap, self).set_options(options_list)
+        self.service_url = options_list['service_url'].get_value()
+        self.verbosity_level = options_list['verbosity_level'].get_value()
+
+    def get_long_desc(self):
+        return (
+            "This plugin parses SOAP according to provided WSDL specification."
+            "User should specify the WSDL URL."
+        )
+
+    def parse_wsdl(self, wsdl_url):
+        """
+        :param str wsdl_url: string address pointing to wsdl spec.
+        :return list: all urls discovered when parsing the wsdl.
+        """
+        discovered_urls = set()
+        try:
+            wsdl_client = zeep.Client(wsdl_url)
+        except XMLSyntaxError:
+            exception_description = (
+                "The result of url: {} seems not to be valid XML.".format(wsdl_url)
+            )
+            self._report_parsing_failure(exception_description, wsdl_url)
+            return discovered_urls
+
+        if not wsdl_client.wsdl.services:
+            exception_description = (
+                "The result of url: {} seems not to be valid WSDL".format(wsdl_url)
+            )
+            self._report_parsing_failure(exception_description, wsdl_url)
+            return discovered_urls
+
+        if self.verbosity_level == 2:
+            dump_capturer = StringIO()
+            with redirect_stdout(dump_capturer):
+                wsdl_client.wsdl.dump()
+            self._report_wsdl_dump(dump_capturer.getvalue())
+
+        for service_name in wsdl_client.wsdl.services:
+            service = wsdl_client.bind(service_name)
+            ports = wsdl_client.wsdl.services[service_name].ports
+            for _, port in ports.items():
+                operation_url = port.binding_options['address']
+                if self.verbosity_level == 1 and operation_url not in discovered_urls:
+                    self._report_all_operations_for_url(service, operation_url)
+                discovered_urls.add(operation_url)
```

Comment:
What about HTTP request parameters? I don't see any handling of POST-body nor query string parameters.

Reply:
I recommend moving the parsing of WSDL to the `parsers` path, where all the other parsers live. The WSDL parser will eventually grow and we don't want it here in the plugin.

I'm not 100% sure that we need this plugin. If you add the WSDL parser to:
https://github.com/andresriancho/w3af/blob/master/w3af/core/data/parsers/document_parser.py#L37

And w3af finds the WSDL URL, it will be able to parse it in `web_spider` and return Fuzzable HTTP requests, which can then be translated into a WSDL mutant, similar to:

https://github.com/andresriancho/w3af/blob/master/w3af/core/data/fuzzer/mutants/json_mutant.py
https://github.com/andresriancho/w3af/blob/master/w3af/core/data/fuzzer/mutants/xmlrpc_mutant.py (this one doesn't work)

The user would just have to add the WSDL to the target URL(s) of w3af, without even having to enable a plugin and set the target there. Also, another benefit of this approach, would be that the user doesn't even know that there is a WSDL endpoint, w3af finds it and analyzes it.

Reply:
`wsdl_parser` created.